org.lflang/src/org/lflang/generator/c/CPortGenerator.java

```diff
@@ -88,22 +96,39 @@ public static String initializeInputMultiport(
     ) {
         var portRefName = CUtil.portRefName(input);
         // If the port is a multiport, create an array.
-        return input.isMultiport() ?
-                String.join("\n",
-                    portRefName+"_width = "+input.getWidth()+";",
-                    "// Allocate memory for multiport inputs.",
-                    portRefName+" = ("+variableStructType(input)+"**)_lf_allocate(",
-                    "        "+input.getWidth()+", sizeof("+variableStructType(input)+"*),",
-                    "        &"+reactorSelfStruct+"->base.allocations); ",
-                    "// Set inputs by default to an always absent default input.",
-                    "for (int i = 0; i < "+input.getWidth()+"; i++) {",
-                    "    "+portRefName+"[i] = &"+reactorSelfStruct+"->_lf_default__"+input.getName()+";",
+        if (input.isMultiport()) {
+            String result = String.join("\n",
+                portRefName+"_width = "+input.getWidth()+";",
+                "// Allocate memory for multiport inputs.",
+                portRefName+" = ("+variableStructType(input)+"**)_lf_allocate(",
+                "        "+input.getWidth()+", sizeof("+variableStructType(input)+"*),",
+                "        &"+reactorSelfStruct+"->base.allocations); ",
+                "// Set inputs by default to an always absent default input.",
+                "for (int i = 0; i < "+input.getWidth()+"; i++) {",
+                "    "+portRefName+"[i] = &"+reactorSelfStruct+"->_lf_default__"+input.getName()+";",
+                "}"
+            );
+            if (AttributeUtils.isSparse(input.getDefinition())) {
+                return String.join("\n", result,
+                    "if ("+input.getWidth()+" >= LF_SPARSE_WIDTH_THRESHOLD) {",
+                    "    "+portRefName+"__sparse = _lf_allocate(1,",
+                    "            sizeof(lf_sparse_io_record_t) + sizeof(size_t) * "+input.getWidth()+"/LF_SPARSE_CAPACITY_DIVIDER,",
+                    "            &"+reactorSelfStruct+"->base.allocations);",
+                    "    "+portRefName+"__sparse->capacity = "+input.getWidth()+"/LF_SPARSE_CAPACITY_DIVIDER;",
+                    "    if (_lf_sparse_io_record_sizes.start == NULL) {",
+                    "        _lf_sparse_io_record_sizes = vector_new(1);",
+                    "    }",
+                    "    vector_push(&_lf_sparse_io_record_sizes, (void*)&"+portRefName+"__sparse->size);",
                     "}"
-                ) :
-                String.join("\n",
-                    "// width of -2 indicates that it is not a multiport.",
-                    portRefName+"_width = -2;"
                 );
+            }
+            return result;
+        } else {
+            return String.join("\n",
+                "// width of -2 indicates that it is not a multiport.",
+                portRefName+"_width = -2;"
```

Comment:
Why not use an `enum` for this and ditch the comment?

Reply:
The width can also be any nonnegative integer. I don't know how to use an enum for that.

Reply:
Ah, right. Never mind this suggestion then.